src/main/java/org/elasticsearch/common/util/MultiDataPathUpgrader.java

```diff
@@ -0,0 +1,366 @@
+                }
+            }
+        }
+        return new ShardPath(target.path, target.path, IndexMetaData.INDEX_UUID_NA_VALUE /* we don't know */, shard);
+    }
+
+    private ShardFileInfo[] getShardFileInfo(ShardId shard, NodeEnvironment.NodePath[] paths) throws IOException {
+        final ShardFileInfo[] info = new ShardFileInfo[paths.length];
+        for (int i = 0; i < info.length; i++) {
+            Path path = paths[i].resolve(shard);
+            final long usabelSpace = getUsabelSpace(paths[i]);
+            info[i] = new ShardFileInfo(path, usabelSpace, getSpaceUsedByShard(path));
+        }
+        return info;
+    }
+
+    protected long getSpaceUsedByShard(Path path) throws IOException {
+        final long[] spaceUsedByShard = new long[] {0};
+        if (Files.exists(path)) {
+            Files.walkFileTree(path, new FileVisitor<Path>() {
+                @Override
+                public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    if (attrs.isRegularFile()) {
+                        spaceUsedByShard[0] += attrs.size();
+                    }
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+        }
+        return spaceUsedByShard[0];
+    }
+
+    protected long getUsabelSpace(NodeEnvironment.NodePath path) throws IOException {
+        FileStore fileStore = path.fileStore;
+        return fileStore.getUsableSpace();
+    }
+
+    static class ShardFileInfo {
+        final Path path;
+        final long usableSpace;
+        final long spaceUsedByShard;
+
+        ShardFileInfo(Path path, long usableSpace, long spaceUsedByShard) {
+            this.path = path;
+            this.usableSpace = usableSpace;
+            this.spaceUsedByShard = spaceUsedByShard;
+        }
+    }
+
+
+
+    private void upgradeFiles(ShardId shard, ShardPath targetPath, final Path targetDir, String folderName, Path[] paths) throws IOException {
+        List<Path> movedFiles = new ArrayList<>();
+        for (Path path : paths) {
+            if (path.equals(targetPath.getDataPath()) == false) {
+                final Path sourceDir = path.resolve(folderName);
+                if (Files.exists(sourceDir)) {
+                    logger.info("{} upgrading {} path {}", shard, folderName, sourceDir);
```

Comment:
can we log "{} upgrading [{}] from [{}] to [{}]", shard, folderName, sourceDir, targetDir" ?